qubesdbus/device_manager.py

```diff
@@ -35,37 +35,218 @@
         level=logging.DEBUG, SYSLOG_IDENTIFIER='qubesdbus.domain_manager'))
 log.propagate = True
 
-pp = pprint.PrettyPrinter()
+SERVICE_NAME = "org.qubes.Devices1"
+SERVICE_PATH = "/org/qubes/Devices1"
+DEV_TYPES = ['block', 'pci', 'usb']  # Add 'mic' when it's implemented
+DEV_IFACE = 'org.qubes.Device'
+
+DBusSignalMatch = dbus.connection.SignalMatch
 
 
 class DeviceManager(qubesdbus.service.ObjectManager):
-    def __init__(self, data):
-        super(DeviceManager, self).__init__()
-        self.managed_objects = [Device(self.bus, self.bus_name, self.bus_path,
-                                       dev) for dev in data]
+    def __init__(self) -> None:
+        super().__init__(SERVICE_NAME, SERVICE_PATH)
+        self.devices = {}  # type: Dict[str, Device]
+
+        for vm in self.app.domains:
+            for dev_class in DEV_TYPES:
+                for dev_info in vm.devices[dev_class].available():
+                    obj_path, device = self._device(vm, dev_class, dev_info)
+                    self.devices[obj_path] = device
+
+        for vm in self.app.domains:
+            for dev_class in DEV_TYPES:
+                for assignment in vm.devices[dev_class].assignments():
+                    try:
+                        obj_path, frontend_vm_path = self._frontend_domain(
+                            vm, assignment, dev_class)
+                        self.devices[obj_path].properties[
+                            'frontend_domain'] = dbus.ObjectPath(
+                                frontend_vm_path)
```

Comment:
frontend_domain makes sense only for running VMs - device can be assigned to multiple VMs, as long as only one of them is running